lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts

```diff
@@ -0,0 +1,302 @@
+import { Currency, CurrencyAmount, Token, TradeType } from '@uniswap/sdk-core'
+import { Protocol } from '@uniswap/router-sdk'
+import { CACHED_ROUTES_CONFIGURATION } from './cached-routes-configuration'
+import { PairTradeTypeChainId } from './model/pair-trade-type-chain-id'
+import { CachedRoutesMarshaller } from './marshalling/cached-routes-marshaller'
+import { CachedRoutesStrategy } from './model/cached-routes-strategy'
+
+interface ConstructorParams {
+  /**
+   * The TableName for the DynamoDB Table. This is wired in from the CDK definition.
+   */
+  cachedRoutesTableName: string
+  /**
+   * The amount of minutes that a CachedRoute should live in the database.
+   * This is used to limit the database growth, and DynamoDB will automatically delete expired entries.
+   */
+  ttl_minutes?: number
+}
+
+export class DynamoRouteCachingProvider extends IRouteCachingProvider {
+  private readonly ddbClient: DynamoDB.DocumentClient
+  private readonly tableName: string
+  private readonly ttl_minutes: number
+
+  constructor({ cachedRoutesTableName, ttl_minutes = 5 }: ConstructorParams) {
+    super()
+    this.ddbClient = new DynamoDB.DocumentClient()
+    this.tableName = cachedRoutesTableName
+    this.ttl_minutes = ttl_minutes
+  }
+
+  /**
+   * Implementation of the abstract method defined in `IRouteCachingProvider`
+   * Given a CachedRoutesStrategy (from CACHED_ROUTES_CONFIGURATION) we will find the BlocksToLive associated to the bucket.
+   *
+   * @param cachedRoutes
+   * @param amount
+   * @protected
+   */
+  protected async _getBlocksToLive(cachedRoutes: CachedRoutes, amount: CurrencyAmount<Currency>): Promise<number> {
+    const cachedRoutesStrategy = this.getCachedRoutesStrategyFromCachedRoutes(cachedRoutes)
+    const cachingParameters = cachedRoutesStrategy?.getCachingParameters(amount)
+
+    if (cachingParameters) {
+      return cachingParameters.blocksToLive
+    } else {
+      return 0
+    }
+  }
+
+  /**
+   * Implementation of the abstract method defined in `IRouteCachingProvider`
+   * Fetch the most recent entry from the DynamoDB table for that pair, tradeType, chainId, protocols and bucket
+   *
+   * @param chainId
+   * @param amount
+   * @param quoteToken
+   * @param tradeType
+   * @param protocols
+   * @protected
+   */
+  protected async _getCachedRoute(
+    chainId: ChainId,
+    amount: CurrencyAmount<Currency>,
+    quoteToken: Token,
+    tradeType: TradeType,
+    protocols: Protocol[]
+  ): Promise<CachedRoutes | undefined> {
+    const [tokenIn, tokenOut] = this.determineTokenInOut(amount, quoteToken, tradeType)
+    const cachedRoutesStrategy = this.getCachedRoutesStrategy(tokenIn, tokenOut, tradeType, chainId)
+    const cachingParameters = cachedRoutesStrategy?.getCachingParameters(amount)
+
+    if (cachingParameters) {
+      const partitionKey = `${tokenIn.address}/${tokenOut.address}/${tradeType}/${chainId}`
+      const partialSortKey = `${protocols}/${cachingParameters.bucket}/`
```

Comment:
Should we serialize the `protocols` array and add a deliminator? Also do you need the trailing `/`?

Reply:
I don't think we need to serialize it, it's an array and we will get something like: V2,V3,MIXED or V3 or V2,V3 and thats what I wanted to get, so protocols is fine, it's mostly a way to have different cache entries depending on the requested protocols.

The trailing `/` is to be able to match the last bit only which is the block number, we could keep it or remove it, but I opted for keeping it to make it obvious that we are just leaving the blockNumber out